### PR TITLE
gitlab modules: deprecate basic auth method

### PR DESCRIPTION
**Pull request by lgatellier**
##### SUMMARY
This PR deprecates basic auth for GitLab modules, since this auth method has been removed from GitLab a **long** time ago (in GitLab 9.5, released in August 2017, if I'm right).

It will allow to remove some specific code from `module_utils/gitlab.py` and enhance maintainability.

##### ISSUE TYPE
- Refactoring Pull Request

##### COMPONENT NAME
gitlab_deploy_key
gitlab_project_badge
gitlab_hook
gitlab_project_variable
gitlab_label
gitlab_instance_variable
gitlab_project_members
gitlab_project_access_token
gitlab_group_access_token
gitlab_merge_request
gitlab_branch
gitlab_project
gitlab_milestone
gitlab_runner
gitlab_protected_branch
gitlab_group_variable
gitlab_user
gitlab_group_members
gitlab_group
gitlab_issue

##### ADDITIONAL INFORMATION
Another PR will follow to remove this code in `community.general` v9.0.0

### CODE REVIEW

**Review by nejch — APPROVED** [review]
LGTM. I seriously doubt there'd still be any GitLab instances out in the wild still supporting basic auth.

Personally I'm also not sure if job token auth makes sense, because none of the current GitLab API resources in the modules here support job token auth. Just for your consideration if you think it would simplify things more. See https://docs.gitlab.com/ee/ci/jobs/ci_job_token.html.

https://github.com/ansible-collections/community.general/blob/cd01a928abad919302e51974fc0c96ececf8ddc0/plugins/module_utils/gitlab.py#L109.

(not sure if this also warrants a changelog entry)

**Comment by lgatellier** [comment]
Thanks for your review @nejch !

> LGTM. I seriously doubt there'd still be any GitLab instances out in the wild still supporting basic auth.

I hope no one still run such old versions. :crossed_fingers: 

> Personally I'm also not sure if job token auth makes sense, because none of the current GitLab API resources in the modules here support job token auth. Just for your consideration if you think it would simplify things more. See https://docs.gitlab.com/ee/ci/jobs/ci_job_token.html.
> 
> https://github.com/ansible-collections/community.general/blob/cd01a928abad919302e51974fc0c96ececf8ddc0/plugins/module_utils/gitlab.py#L109
> .

You're right, but I don't know if it makes sense to remove a feature which would be re-added later for another module (like a gitlab release creation for example).

> (not sure if this also warrants a changelog entry)

Don't know either. I'm also not sure about how to handle sanity issues on `module.deprecate()` call.

A collection maintainer may help us of the last 3 points.

File: changelogs/fragments/8383-deprecate-gitlab-basic-auth.yml
Comment (felixfontein) [inline changelogs/fragments/8383-deprecate-gitlab-basic-auth.yml:2]: The next feature release is 9.0.0, so the removal version needs to be at least one major version later.
```suggestion
  - "gitlab modules - the basic auth method on GitLab API have been deprecated and will be removed in community.general 10.0.0 (https://github.com/ansible-collections/community.general/pull/8383)."
```

File: plugins/module_utils/gitlab.py
Comment (felixfontein) [inline plugins/module_utils/gitlab.py:121]:
```suggestion
                version='10.0.0',
```

File: tests/sanity/ignore-2.13.txt
Comment (felixfontein) [inline tests/sanity/ignore-2.13.txt:4]: Please remove these entries :)

**Review by felixfontein — COMMENTED** [review]
Thanks for your contribution!

File: tests/sanity/ignore-2.13.txt
Comment (felixfontein) [reply tests/sanity/ignore-2.13.txt:4]: (This would have to be a `pylint:collection-deprecated-version` ignore anyway, since it's failing the `pylint` test and not the `validate-modules` test :) )

**Comment by lgatellier** [comment]
@felixfontein I just applied your remarks :wink: 

Should I open another PR removing the deprecated code now, or should I do it once 9.0.0 will be released ?

**Comment by felixfontein** [comment]
> Should I open another PR removing the deprecated code now, or should I do it once 9.0.0 will be released ?

Once this is merged ;) It won't get merged until we have a stable-9 branch, but since that will happen tomorrow, feel free to already create it ;-)

**Comment by felixfontein** [comment]
> > Personally I'm also not sure if job token auth makes sense, because none of the current GitLab API resources in the modules here support job token auth. Just for your consideration if you think it would simplify things more. See https://docs.gitlab.com/ee/ci/jobs/ci_job_token.html.
> > https://github.com/ansible-collections/community.general/blob/cd01a928abad919302e51974fc0c96ececf8ddc0/plugins/module_utils/gitlab.py#L109
>
> You're right, but I don't know if it makes sense to remove a feature which would be re-added later for another module (like a gitlab release creation for example).

That's a good question. It would probably be a good idea to move it to a separate docs fragment, and modify the argspec creation in module_utils to add this only when requested.

Then we can deprecate it from the current set of modules, but still use it later for specific modules where this is actually needed.

What do you think?

**Review by felixfontein — APPROVED** [review]

**Comment by felixfontein** [comment]
@lgatellier thanks for your contribution!
@nejch thanks for reviewing!

**Comment by lgatellier** [comment]
> That's a good question. It would probably be a good idea to move it to a separate docs fragment, and modify the argspec creation in module_utils to add this only when requested.
> 
> Then we can deprecate it from the current set of modules, but still use it later for specific modules where this is actually needed.
> 
> What do you think?

Thats a nice solution, thanks for your advice !

**Pull request merged**
